front/src/modules/trainschedule/components/ImportTrainSchedule/ImportTrainScheduleConfig.tsx

```diff
@@ -144,17 +213,235 @@ const ImportTrainScheduleConfig = ({
+
+        ocps.forEach((ocp) => {
+          const id = ocp.getAttribute('id');
+          const code = ocp.getAttribute('code');
+
+          if (id && code) {
+            const { ciCode, chCode } = extractCiChCode(code);
+            localCichDict[id] = { ciCode, chCode };
+          }
+        });
+      });
+    });
+
+    // Ensure cichDict is fully populated before proceeding
+    localCichDict = await overloadDict(localCichDict);
+
+    const trainParts = Array.from(xmlDoc.getElementsByTagName('trainPart'));
+    const period = xmlDoc.getElementsByTagName('timetablePeriod')[0];
+    const startDate = period ? period.getAttribute('startDate') : null;
+
+    if (!startDate) {
+      console.error('Start Date not found in the timetablePeriod.');
+      return trainSchedules;
+    }
+
+    trainParts.forEach((train) => {
+      const trainNumber = train.getAttribute('id') || '';
+      const ocpSteps = Array.from(train.getElementsByTagName('ocpTT'));
+      const formationTT = train.getElementsByTagName('formationTT')[0];
+      const rollingStockViriato = formationTT?.getAttribute('formationRef');
+      const firstOcpTT = ocpSteps[0];
+      const firstDepartureTime = firstOcpTT
+        .getElementsByTagName('times')[0]
+        ?.getAttribute('departure');
+
+      const firstDepartureTimeformatted = firstDepartureTime && cleanTimeFormat(firstDepartureTime);
+
+      const lastOcpTT = ocpSteps[ocpSteps.length - 1];
+      const lastDepartureTime =
+        lastOcpTT.getElementsByTagName('times')[0]?.getAttribute('departure') ||
+        lastOcpTT.getElementsByTagName('times')[0]?.getAttribute('arrival');
+      const lastDepartureTimeformatted = lastDepartureTime && cleanTimeFormat(lastDepartureTime);
+
+      // Build steps using the fully populated localCichDict
+      const adaptedSteps = buildSteps(ocpSteps, localCichDict, startDate);
+
+      const trainSchedule: ImportedTrainSchedule = {
+        trainNumber,
+        rollingStock: rollingStockViriato, // RollingStocks in viriato files rarely have the correct format
+        departureTime: `${startDate} ${firstDepartureTimeformatted}`,
+        arrivalTime: `${startDate} ${lastDepartureTimeformatted}`,
+        departure: '', // Default for testing
+        steps: adaptedSteps,
+      };
+
+      trainSchedules.push(trainSchedule);
+    });
+    const trains = Array.from(xmlDoc.getElementsByTagName('train'));
+    const updatedTrainSchedules = mapTrainNames(trainSchedules, trains);
+
+    return updatedTrainSchedules;
+  };
 
   const importFile = async (file: File) => {
     closeModal();
     setTrainsList([]);
 
-    const text = await file.text();
-    const importedTrainSchedules: TrainScheduleBase[] = JSON.parse(text);
+    const fileName = file.name.toLowerCase();
+    const fileExtension = fileName.split('.').pop();
 
-    if (!isEmpty(importedTrainSchedules)) {
-      setTrainsJsonData(importedTrainSchedules);
-    }
+    const reader = new FileReader();
```

Comment:
Could probably use `await file.text()` like we do in the validation function instead of `FileReader`?